### PR TITLE
Add support for primitive Optional classes #379

### PR DESCRIPTION
**Pull request by AlexBischof**
This PR contains an extension of the API for OptionalInt, OptionalLong and OptionalDouble as described in the corresponding issue. It contains tests and documentation and I have oriented myself on existing code. I also extended BDDAssertions as well as Softassertions for those classes.

Here are some examples:
```java
assertThat(OptionalDouble.of(8)).hasValueCloseTo(8.0, within(0d));
assertThat(OptionalDouble.of(8)).hasValue(8.0);

assertThat(OptionalInt.of(7)).hasValue(7);
assertThat(OptionalLong.of(7)).hasValue(7);

assertThat(OptionalInt.empty()).isEmpty();
```

One thing seems not right, yet. The generic class Optional contains a value whereas an OptionalXXX has a value. 
```java
assertThat(Optional.of("Hello World")).contains("Hello World");
assertThat(OptionalDouble.of(8)).hasValue(8.0);
```
It is not konsistent, is it?

Do you need more help on that?

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractOptionalIntAssert.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     * @throws AssertionError if actual value is empty.
+     * @throws AssertionError if actual is null.
+     * @throws AssertionError if actual has not the value as expected.
+     */
+    public S hasValue(int expectedValue) {
+        isNotNull();
+        if (!actual.isPresent()) throw failure(shouldContain(expectedValue));
+        //Reuses integers functionality, catches possible assertionerror and throw correct one
+        try {
+            integers.assertEqual(info, actual.getAsInt(), expectedValue);
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractOptionalIntAssert.java]: In that case, just comparing `int` is easier:
```java
if (expectedInt != actual.getAsInt()) throw Failures.instance().failure(info, shouldContain(actual, expectedValue));
```

File: src/main/java/org/assertj/core/api/AbstractOptionalLongAssert.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * Assertions for {@link java.util.OptionalLong}.
+ *
+ * @author Jean-Christophe Gay
+ * @author Alexander Bischof
+ */
+public abstract class AbstractOptionalLongAssert<S extends AbstractOptionalLongAssert<S>> extends
+    AbstractAssert<S, OptionalLong> {
+
+    @VisibleForTesting
+    Integers integers = Integers.instance();
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractOptionalLongAssert.java]: `Integers` in `AbstractOptionalLongAssert` ??? 
Anyway, we don't need that as per my previous comment.

File: src/main/java/org/assertj/core/error/OptionalShouldBeEmpty.java [hunk 3]
```diff
@@ -23,18 +24,48 @@
  */
 public class OptionalShouldBeEmpty extends BasicErrorMessageFactory {
 
-  private OptionalShouldBeEmpty(Object optionalValue) {
-	super("%nExpecting an empty Optional but was containing value: <%s>.", optionalValue);
-  }
+    private OptionalShouldBeEmpty(Object optionalValue) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/OptionalShouldBeEmpty.java]: add a `String` parameter to get the `Optional` type to have more precise error message:
```
"Expecting an empty %s but was containing value: <%s>."
```
That should give for `OptionalDouble`: 
```
"Expecting an empty OptionalDouble but was containing value: <1.0>."
```

File: src/main/java/org/assertj/core/error/OptionalShouldContain.java [hunk 4]
```diff
@@ -12,44 +12,78 @@
...
-  /**
-   * Indicates that an {@link java.util.Optional} is empty so it doesn't contain the expected value.
-   *
-   * @param expectedValue the value we expect to be in an {@link java.util.Optional}.
-   * @return a error message factory.
-   */
-  public static OptionalShouldContain shouldContain(Object expectedValue) {
-	return new OptionalShouldContain(expectedValue);
-  }
+    private OptionalShouldContain(Object actual, Object expected) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/OptionalShouldContain.java]: Similar comment as `OptionalShouldBeEmpty` to add optional type in the message

**Comment by AlexBischof** [comment]
I fixed the comparison accordingly (int, long, double) and improved the OptionalShouldXXX classes so that the actual optional class appears in the error message.

**Comment by joel-costigliola** [comment]
cool, thanks !

**Comment by joel-costigliola** [comment]
I just read your comment on `hasValue` vs `contains`, it is indeed inconsistent but I think `hasValue`is a good name which works well wth `.hasValueCloseTo(8.0, within(0d));` whereas `.containsCloseTo(8.0, within(0d));` is a little bit weird.

We can't replace `contains` by `hasValue` in optional assertions, it would break the API.
If someone requests it, we might just add `hasValue` as an alias of `contains`.

**Comment by joel-costigliola** [comment]
Merged, thanks.

**Pull request closed without merging**